servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -441,54 +462,125 @@ public Completable closeAsyncGracefully() {
+                        .flatMapCompletable(newCnx -> {
+                            if (!selector.test(newCnx)) {
+                                return newCnx.closeAsync().concat(Completable.failed(new RuntimeException()));
+                            }
+                            if (host.addConnection(newCnx)) {
+                                host.markHealthy();
+                            } else {
+                                return newCnx.closeAsync().concat(Completable.failed(new RuntimeException()));
+                            }
+                            return completed();
+                        })
+                        .afterOnError(e -> {
+                            if (!cancelled) {
+                                schedule();
+                            }
+                        })
+                        .subscribe();
+            }
+
+            @Override
+            public void cancel() {
+                this.cancelled = true;
+            }
         }
 
         private static final class ConnState {
             final Object[] connections;
-            final State state;
+            final Object state;
 
-            ConnState(final Object[] connections, final State state) {
+            ConnState(final Object[] connections, final Object state) {
                 this.connections = connections;
                 this.state = state;
             }
         }
 
         final Addr address;
+        @Nullable
+        private final HealthCheckConfig healthCheckConfig;
         private final ListenableAsyncCloseable closeable;
-
         private volatile ConnState connState = EMPTY_CONN_STATE;
+        private volatile int connectionFailures;
 
-        Host(Addr address) {
+        Host(Addr address, @Nullable HealthCheckConfig healthCheckConfig) {
             this.address = requireNonNull(address);
+            this.healthCheckConfig = healthCheckConfig;
             this.closeable = toAsyncCloseable(graceful ->
                     graceful ? doClose(AsyncCloseable::closeAsyncGracefully) : doClose(AsyncCloseable::closeAsync));
         }
 
-        boolean tryToMarkActive() {
-            return connStateUpdater.updateAndGet(this, oldConnState -> {
-                if (oldConnState.state == State.EXPIRED) {
-                    return new ConnState(oldConnState.connections, State.ACTIVE);
+        boolean markActiveIfNotClosed() {
+            final Object oldState = connStateUpdater.getAndUpdate(this, oldConnState -> {
+                if (oldConnState.state == STATE_EXPIRED) {
+                    return new ConnState(oldConnState.connections, STATE_ACTIVE);
                 }
                 // If oldConnState.state == State.ACTIVE this could mean either a duplicate event,
                 // or a repeated CAS operation. We could issue a warning, but as we don't know, we don't log anything.
+                // UNHEALTHY state is treated similarly to ACTIVE
```

Comment:
This comment is a bit confusing. Consider saying: UNHEALTHY state cannot transition to ACTIVE without passing the health check.